File: usaco/training/section_3.3/camelot.py
"""
ID: huangch7
LANG: PYTHON3
TASK: camelot
"""


task_name = "camelot"
INF = 10000000000

def get_ints_from_line():
    return [int(x) for x in fin.readline().strip().split()]

def get_strs_from_line():
    return fin.readline().strip().split()

fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

R, C = get_ints_from_line()

pos = []

while True:
    cols = get_strs_from_line()
    if len(cols) == 0:
        break
    for i in range(0, len(cols), 2):
        pos.append((int(cols[i + 1]) - 1, ord(cols[i]) - ord('A')))

king = pos[0]
knights = pos[1:]

# print(king)
# print(len(knights))

paths = {}

def dp(x1, y1, x2, y2):

    if (x1, y1, x2, y2) in paths:
        return paths[(x1, y1, x2, y2)]
    # position, distance, pick_king
    deque = [((x1, y1), 0)]
    paths[(x1, y1, x1, y1)] = 0
    while len(deque):
        start, dist = deque.pop(0)
        for x, y in [(-1, 2), (-1, -2), (1, -2), (1, 2),
                     (-2, 1), (-2, -1), (2, -1), (2, 1)]:
            x_mid, y_mid = start[0] + x, start[1] + y
            if x_mid < 0 or x_mid >= R or y_mid < 0 or y_mid >= C:
                continue
            if (x1, y1, x_mid, y_mid) in paths:
                continue
            paths[(x1, y1, x_mid, y_mid)] = 1 + dist
            paths[(x_mid, y_mid, x1, y1)] = 1 + dist
# TODO: if add this, WA
#             if (x1, y1, x2, y2) in paths:
#                 break
            deque.append(((x_mid, y_mid), dist + 1))
    if (x1, y1, x2, y2) not in paths:
        paths[(x1, y1, x2, y2)] = INF
        paths[(x2, y2, x1, y1)] = INF

    return paths[(x1, y1, x2, y2)]


# print('debug-1', dp(2, 0, 2, 3))
# print('debug-1', dp(2, 3, 2, 0))


best = None
for x in range(R):
    for y in range(C):
        total = 0

        for x_k, y_k in knights:
            total += dp(x_k, y_k, x, y)

        total_knights = total
        total += max(abs(x - king[0]), abs(y - king[1]))
        if best is None or best > total:
            best = total
        for x_k, y_k in knights:
            pick_king_dist = total_knights - dp(x_k, y_k, x, y)
            pick_king_dist += dp(x_k, y_k, king[0], king[1])
            pick_king_dist += dp(king[0], king[1], x, y)
            best = min(best, pick_king_dist)
#            if x == 2 and y == 3:
#                print('debug', x_k, y_k, total_knights, dp(x_k, y_k, x, y), dp(x_k, y_k, king[0], king[1]), dp(king[0], king[1], x, y))
#         print(x, y, total_knights, best)

fout.write(f"{best}\n")

# print(dp(2, 0, 4, 1))
# print(pick_king[(2, 0, 4, 1)])
# print(dp(7, 0, 4, 1))
# print(pick_king[(7, 0, 4, 1)])
# print(dp(0, 7, 4, 1))
# print(pick_king[(0, 7, 4, 1)])
# print(dp(7, 7, 4, 1))
# print(pick_king[(7, 7, 4, 1)])
    
fout.close()

File: usaco/training/section_3.3/camelot.cpp
/*
ID: huangch7
LANG: C++
TASK: camelot
*/

#include <iostream>
#include <fstream>
#include <deque>
#include <unordered_map>
#include <cmath>
#include <cstring>

using namespace std;

const int INF = 100000;

int R, C;
int king_x, king_y;
int knight_pos[30][30];

int dist_cache[30][30][30][30];

 


int king_distance(int x1, int y1, int x2, int y2) {
    int dx = abs(x1 - x2);
    int dy = abs(y1 - y2);
    return max(dx, dy);
}



int bfs(int sx, int sy, int ex, int ey) {
  if (dist_cache[sx][sy][ex][ey] != -1) {
    return dist_cache[sx][sy][ex][ey];
  }
    bool visited[30][30] = {false};
    int dist[30][30] = {0};
    int dx[] = {-1, -1, 1, 1, -2, -2, 2, 2};
    int dy[] = {2, -2, 2, -2, 1, -1, 1, -1};
    
    deque<pair<int, int> > q;
    q.push_back(make_pair(sx, sy));
    visited[sx][sy] = true;
    dist_cache[sx][sy][sx][sy] = 0;
    
    while (!q.empty()) {
        pair<int, int> curr = q.front();
        q.pop_front();
        int x = curr.first;
        int y = curr.second;
        
//        if (x == ex && y == ey) {
//            return dist[x][y];
//        }
//        
        for (int i = 0; i < 8; ++i) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if (nx >= 0 && nx < R && ny >= 0 && ny < C && !visited[nx][ny]) {
                visited[nx][ny] = true;
                dist[nx][ny] = dist[x][y] + 1;
                dist_cache[sx][sy][nx][ny] = dist[x][y] + 1;
                dist_cache[nx][ny][sx][sy] = dist[x][y] + 1;
                q.push_back(make_pair(nx, ny));
            }
        }
    }
    if (dist_cache[sx][sy][ex][ey] == -1) {
      dist_cache[sx][sy][ex][ey] = INF;
      dist_cache[ex][ey][sx][sy] = INF;
    }
    
    return dist_cache[sx][sy][ex][ey];
}

int main() {
    ifstream fin("camelot.in");
    ofstream fout("camelot.out");
 for(int i = 0; i < 30; ++i)
   for(int j = 0; j < 30; ++j)
     for(int k = 0; k < 30; ++k)
        for(int p = 0; p < 30; ++p)
          dist_cache[i][j][k][p] = -1;
    

    fin >> R >> C;
    char col;
    int row;
    fin >> col >> row;
    king_x = row - 1;
    king_y = col - 'A';
    
    while (fin >> col >> row) {
        knight_pos[row - 1][col - 'A'] = 1;
    }

    int min_moves = INF;
    for (int i = 0; i < R; ++i) {
        for (int j = 0; j < C; ++j) {
            int total_dist = 0;
            for (int k = 0; k < R; ++k) {
                for (int l = 0; l < C; ++l) {
                    if (knight_pos[k][l]) {
                        total_dist += bfs(k, l, i, j);
                    }
                }
            }
            //            int dist_king = king_distance(i, j, king_x, king_y);
            //            int total = total_dist + dist_king;
            int best = INF;
            for (int len = 0; len < max(R, C); ++len){
              for (int px = max(0, king_x - len); px < min(R, king_x + len + 1); ++px) {
                 for (int py = max(0, king_y - len); py < min(C, king_y + len + 1); ++py) {
                    for (int k = 0; k < R; ++k) {
                       for (int l = 0; l < C; ++l) {
                         if (knight_pos[k][l]) {
                   
                            if (bfs(k, l, px, py) + bfs(px, py, i, j) == bfs(k, l, i, j)){
                              best = len;
                              break;
                            }
                        }
                      }
                       if (best != INF) break;
                    }
                    if (best != INF) break;
                 }
                }
                if (best != INF) break;
                 
            }
            min_moves = min(min_moves, total_dist + min(best, king_distance(i, j, king_x, king_y)));

//            for (int k = 0; k < R; ++k) {
//                for (int l = 0; l < C; ++l) {
//                    if (knight_pos[k][l]) {
//                      // pick up position
//                      for (int px = 0; px < R; ++px) for (int py = 0; py < C; ++ py) {
//                        
//                          int pick_king = total_dist - bfs(k, l, i, j) + bfs(k, l, px, py) + bfs(px, py, i, j) + king_distance(px, py, king_x, king_y);
//                        // int pick_king = total_dist - bfs(k, l, i, j) + bfs(k, l, king_x, king_y) + bfs(king_x, king_y, i, j);
//                        total = min(total, pick_king);
//                      }
//                    }
//                }
//            }
//            min_moves = min(min_moves, total);
        }
    }

    fout << min_moves << endl;

    fin.close();
    fout.close();

    return 0;
}
